File: skills/transparencia-es/scripts/buscar_contratos.ts
// buscar_contratos.ts — implementación de referencia para la edge function de
// Lovable Cloud (Supabase Edge Functions, Deno/TypeScript).
//
// La edge function "chat" declara esta función como TOOL para Lovable AI (Gemini,
// function calling). Replica la lógica de buscar_contratos.py: lee el dataset
// normalizado (contratos_demo.json en Supabase Storage o generado con etl_placsp.py)
// y filtra por órgano / CPV / año. Mantiene el contrato de datos congelado.
//
// Esquema de cada contrato:
//   { objeto, importe, adjudicatario, cpv, tipo_contrato, fecha, estado,
//     organo, expediente, url_expediente }

export interface Contrato {
  objeto: string;
  importe: number | null;
  adjudicatario: string;
  cpv: string;
  tipo_contrato: string;
  fecha: string;
  estado: string;
  organo: string;
  expediente: string;
  url_expediente: string;
}

export interface BuscarArgs {
  organo?: string;
  cpv?: string;
  anio?: string;
}

// Procedencia del dataset (la _meta del JSON). La UI la usa para decidir si
// muestra el aviso de demo: si es_demo===true / aviso_ui!=null => banner.
export interface Fuente {
  naturaleza: string | null;
  es_demo: boolean;
  aviso_ui: string | null;
  fecha_cache: string | null;
  fuente_oficial: string | string[] | null;
}

// Forma del JSON en Storage: { _meta, contratos: [...] }
export interface Dataset {
  _meta?: Partial<Fuente> & Record<string, unknown>;
  contratos: Contrato[];
}

export interface BuscarResult {
  contratos: Contrato[];
  n: number;
  total_importe: number;
  fuente: Fuente;
}

function matches(rec: Contrato, { organo, cpv, anio }: BuscarArgs): boolean {
  if (organo && !(rec.organo ?? "").toLowerCase().includes(organo.toLowerCase())) {
    return false;
  }
  if (cpv && !(rec.cpv ?? "").startsWith(cpv)) return false;
  if (anio && !(rec.fecha ?? "").startsWith(anio)) return false;
  return true;
}

/**
 * Filtra el dataset de contratos. `dataset` es el JSON completo cargado desde
 * Supabase Storage (con `_meta` y `contratos`); cárgalo una vez y cachéalo en
 * la edge function. Acepta también el array `contratos` suelto por comodidad.
 */
export function buscarContratos(
  dataset: Dataset | Contrato[],
  args: BuscarArgs,
): BuscarResult {
  const contratos = Array.isArray(dataset) ? dataset : dataset.contratos;
  const meta = Array.isArray(dataset) ? {} : (dataset._meta ?? {});
  const hits = contratos.filter((rec) => matches(rec, args));
  const total = Math.round(
    hits.reduce((acc, r) => acc + (r.importe ?? 0), 0) * 100,
  ) / 100;
  return {
    contratos: hits,
    n: hits.length,
    total_importe: total,
    fuente: {
      naturaleza: (meta.naturaleza as string) ?? null,
      es_demo: (meta.es_demo as boolean) ?? true,
      aviso_ui: (meta.aviso_ui as string) ?? null,
      fecha_cache: (meta.fecha_cache as string) ?? null,
      fuente_oficial: (meta.fuente_oficial as string | string[]) ?? null,
    },
  };
}

// Definición de la TOOL para Lovable AI / Gemini (function calling):
export const buscarContratosTool = {
  name: "buscar_contratos",
  description:
    "Busca contratos públicos reales (PLACSP) por órgano de contratación, " +
    "código CPV (prefijo) y/o año. Devuelve la lista normalizada y el importe total.",
  parameters: {
    type: "object",
    properties: {
      organo: { type: "string", description: "Nombre o parte del nombre del órgano (ej. 'Ayuntamiento de Barcelona')" },
      cpv: { type: "string", description: "Código o prefijo CPV (ej. '4521' para obras educativas)" },
      anio: { type: "string", description: "Año en formato AAAA" },
    },
    required: ["organo"],
  },
} as const;

File: skills/transparencia-es/scripts/buscar_contratos.py
#!/usr/bin/env python3
"""
buscar_contratos.py — consulta el dataset normalizado de contratos PLACSP.

Es la herramienta que la skill invoca en local (Claude Code / Cursor). En el
chatbot web, la edge function de Lovable replica esta misma lógica de filtrado
sobre el JSON guardado en Storage (ver buscar_contratos.ts).

Lee por defecto data/contratos_demo.json (muestra) o el que generes con
etl_placsp.py, filtra por órgano / CPV / año y devuelve JSON con la forma:

    { "contratos": [ {objeto, importe, adjudicatario, cpv, tipo_contrato,
                      fecha, estado, organo, expediente, url_expediente}, ... ],
      "total_importe": <suma>, "n": <num> }

Ejemplos:
    python buscar_contratos.py --organo "Barcelona" --cpv 4521
    python buscar_contratos.py --organo "Barcelona" --anio 2024 \\
        --dataset ../data/contratos_demo.json
"""

from __future__ import annotations

import argparse
import json
import os
import sys

DEFAULT_DATASET = os.path.join(
    os.path.dirname(__file__), "..", "data", "contratos_demo.json"
)


def matches(rec, organo, cpv, anio):
    if organo and organo.lower() not in (rec.get("organo") or "").lower():
        return False
    if cpv and not (rec.get("cpv") or "").startswith(cpv):
        return False
    if anio and not (rec.get("fecha") or "").startswith(anio):
        return False
    return True


def buscar(dataset_path, organo=None, cpv=None, anio=None):
    with open(dataset_path, encoding="utf-8") as fh:
        data = json.load(fh)
    contratos = data.get("contratos", data if isinstance(data, list) else [])
    hits = [r for r in contratos if matches(r, organo, cpv, anio)]
    total = round(sum(r.get("importe") or 0 for r in hits), 2)
    return {
        "contratos": hits,
        "n": len(hits),
        "total_importe": total,
        "_fuente": data.get("_meta", {}),
    }


def main(argv=None):
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--organo", help="Subcadena del nombre del órgano")
    ap.add_argument("--cpv", help="Prefijo de CPV (p.ej. 4521)")
    ap.add_argument("--anio", help="Año (AAAA)")
    ap.add_argument("--dataset", default=DEFAULT_DATASET, help="Ruta del JSON")
    args = ap.parse_args(argv)

    if not os.path.exists(args.dataset):
        raise SystemExit(f"No existe el dataset: {args.dataset}")

    result = buscar(args.dataset, args.organo, args.cpv, args.anio)
    print(json.dumps(result, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()
